apps/node/server.js

```diff
@@ -0,0 +1,160 @@
+const express = require("express");
+const axios = require("axios");
+const dotenv = require("dotenv").config();
+const crypto = require("crypto");
+const qs = require("qs");
+
+const { getState, setState } = require("./dal");
+
+const ATTENTIVE_API_URL = "https://api-devel.attentivemobile.com/v1";
+const ACCESS_TOKEN_ENDPOINT =
+  "https://api-devel.attentivemobile.com/v1/authorization-codes/tokens";
+
+const CLIENT_ID = dotenv.parsed.CLIENT_ID; // Your client id
+const CLIENT_SECRET = dotenv.parsed.CLIENT_SECRET; // Your secret
+const REDIRECT_URI = dotenv.parsed.REDIRECT_URI; // Your redirect uri
+const PORT = 3002;
+
+let app = express();
+app.use(express.json());
+
+app.get("/", (req, res) => {
+  try {
+    res.json({ message: "Your Application is running successfully!" });
+  } catch (ex) {
+    console.error("unexpected error", ex);
+    res.status(500).json({ message: "Unexpected error" });
+  }
+});
+
+app.get("/install", async (req, res) => {
+  try {
+    // Generates a random string containing numbers and letters
+    const state = crypto.randomBytes(5).toString("hex");
+
+    // List of scopes that your app needs
+    const scopes = ["ecommerce:write", "events:write", "subscriptions:write"];
+
+    setState(state, {});
+
+    // build redirect url and return 302
+    const redirect_url =
+      `https://ui-devel.attentivemobile.com/integrations/oauth-install?client_id=${CLIENT_ID}` +
+      `&redirect_uri=${REDIRECT_URI}` +
+      `&state=${state}`;
+    res.status(302).redirect(redirect_url);
+  } catch (ex) {
+    console.error("unexpected error", ex);
+    res.status(500).json({ message: "Unexpected error" });
+  }
+});
+
+app.get("/callback", async (req, res) => {
+  try {
+    const authorization_code = req.query.code;
+    const state = req.query.state;
+
+    let application = getState(state);
+    if (!application) {
+      return res.status(400).send({
+        message: `Your Application did not successfully install due to no application in memory`,
+      });
+    }
+
+    const payload = {
+      grant_type: "authorization_code",
+      code: authorization_code,
+      redirect_uri: REDIRECT_URI,
+      client_id: CLIENT_ID,
+      client_secret: CLIENT_SECRET,
+    };
+    // your application makes a request to exchange an authorization code for an access_token
+    const response = await axios.post(
+      ACCESS_TOKEN_ENDPOINT,
+      qs.stringify(payload),
+      {
+        headers: {
+          "Content-Type": "application/x-www-form-urlencoded",
+        },
+      }
+    );
+    if (!response) {
```

Comment:
I'm pretty sure axios won't give you an undefined response. It'll throw if the request fails in any way so you may not need these checks (if you were using fetch you would since it doesn't throw).

Reply:
removed!